4단계 - 코드리뷰 요청드립니다.

grow-with opened the pull request:
안녕하세요
3단계에서 리뷰해주신 피드백을 보고 많이 고민을 해보았습니다.
기존에 제가 만들었던 클래스들이 책임을 너무 잘게 쪼개서 분할이 된건가 하는 생각이 들었습니다. 
그래서 책임을 기준으로 생각해보았습니다.
화면에 출력하는 값들은 대부분 서식이 정해져있는데 도메인 클래스에서 대부분 해당 값들을 만들어서 넘겨주는 형태로 코드들이 구현되어 있는것 같다는 생각이 들었습니다.
그래서 아예 도메인에서는 로또 자체를 넘기고 view에서 필요한 정보들을 추출해서 서식에 맞게 가공하는 형태로 분할을 해보았습니다.
그리고 입력받는 과정에서 검증도 도메인로직에서 이루어지고 있는데 공통된 부분을 제외하고는 입력값에 대한 검증은 view에서 처리하도록 수정해보았습니다.
입출력에 대한 책임은 view에게 상당부분을 넘겨보니 클래스가 많이 줄어들었습니다.
기존에 통계를 위해 사용했던 클래스들은 아에 제거 할 수 있었습니다.
또 주신 피드백에 따라서 LottoShop을 만들어서 아예 Lotto 클래스 자체를 반환하도록 수정해보았습니다.
수정이 되면서 새로 테스트코드들 또한 작성해보았고, 알려주신 불변객체로 최대한 유지할 수 있도록 고민해보면서 코드를 작성했습니다.
어제 밤부터 했는데 아침이 되서야 끝나네요 ㅠ 아직 갈길이 먼것 같습니다.
부족한 부분이 있다면 피드백 부탁드리겠습니다. 
감사합니다~

grow-with committed:
refactor: 도메인 로직에서 view 서식 제거

grow-with committed:
refactor: LottoNumber 클래스 생성 및 전체 수정

- LottoShop 클래스 생성 (기존의 factory클래스 제거)
- 각 도메인 클래스 마다 테스트코드 신규로 작성
- 추후 점진적으로 클래스 합치거나 제거

grow-with committed:
docs: 기능 요구사항 수정

grow-with committed:
feat: 보너스 볼 입력받기 및 통계에 추가

- 기능 추가 외에 전체적으로 구조 수정
- 화면 입력받는 값들에 대한 검증 view로 전부 이관
- 기존의 통계관련 클래스들 전부 삭제
- 통계값은 화면에 종속적인 서식을 가지므로 view에서 처리

grow-with committed:
docs: 기능 요구사항 수정

leeyohan93 commented on src/main/java/lotto/controller/LottoApplicationController.java:
```diff
@@ -4,61 +4,42 @@
-		}
-		return "";
-	}
+	public static final int LOTTO_PRICE = 1000;
```
로또 가격에 대한 상수는 Lotto 클래스에 위치하는 것이 더욱 적합할 수 있지 않을까요 ? :)

grow-with replied:
저도 고민했던 내용이긴한데 ㅠ
실제 Lotto 클래스에서 사용하지 않더라도 좀더 의미에 부합하는 곳에 상수를 선언하는 것이 좋을까요??

leeyohan93 replied:
충분히 고민이 되는 좋은 포인트네요! 👍
그렇다면 LottoShop에 배치하게 된다면 상수를 사용하면서 적절하게 배치할 수도 있을 것 같은데 어떠신가요 ? 🤔

leeyohan93 commented on src/main/java/lotto/controller/LottoApplicationController.java:
```diff
@@ -4,61 +4,42 @@
-	public int getPurchaseQuantity() {
-		return purchaseAmount.purchase();
+	public int getPurchaseQuantity(int purchaseAmount) {
+		return purchaseAmount / LOTTO_PRICE;
```
원시 값을 포장하는 Money와 같은 클래스를 도출해보는 것은 어떨까요 ? 
Rank의 amount나 상수로 관리하던 LOTTO_PRICE와 같은 금액을 의미했던 int 타입들을 포장할 수 있을 것 같아요!
Money 클래스는 어떤 적합한 책임을 할당할 수 있을 지 고민해보시면 좋을 것 같아요 👍 


> 규칙 3: 모든 원시값과 문자열을 포장한다.

```suggestion
	public int getPurchaseQuantity(Money purchaseAmount) {
		return purchaseAmount.divide(LOTTO_PRICE);
```

leeyohan93 commented on src/main/java/lotto/controller/LottoApplicationController.java:
```diff
@@ -4,61 +4,42 @@
+			lottos.add(LottoShop.sell());
 			purchaseQuantity--;
 		}
+		this.lottos = new Lottos(lottos);
```
IntStream을 사용해보시는건 어떨까요 ? 
더욱 직관적인 코드가 될 수 있을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/domain/LottoShop.java:
```diff
@@ -0,0 +1,26 @@
+
+	public static Lotto sell(String lottoNumbers) {
+		return new Lotto(new LottoNumbers(lottoNumbers));
+	}
```
로또의 생성과 판매를 담당하는 LottoShop 클래스 도출 하셨네요 💯 

sell, `판매하다` 라는 메서드 명을 사용해주셨는데, 지금은 생성해주는 역할만 하고 있네요.
메서드의 인자로 금액을 의미하는 메서드 인자를 설정하고, 
메서드 구현에서 금액만큼 로또를 발급한다면 더욱 적합한 책임을 할 수 있지 않을까요? :)

leeyohan93 commented on src/main/java/lotto/domain/Repeater.java:
```diff
@@ -14,7 +14,11 @@ public static boolean isContinue() {
 		return repeater;
 	}
 
```
Repeater 클래스는 콘솔 화면에 대한 요구사항들을 처리하고 있는데요 !
도메인 패키지보다는 적합한 패키지가 있을 것 같아요 🤔

leeyohan93 commented on src/main/java/lotto/controller/LottoApplicationController.java:
```diff
@@ -4,61 +4,42 @@
+	public List<Lotto> recorde(String winningNumbers, int bonusBallNumber) {
+		for (Lotto lotto : lottos.getLottos()) {
+			lotto.recordeRank(new LottoNumbers(winningNumbers), new LottoNumber(bonusBallNumber));
 		}
```
일급 컬렉션인 Lottos에게 책임을 이전해보면 어떨까요.
일급 컬렉션의 getter는 어쩔 수 없이 뷰를 구현하기 위한 데이터를 전달해 줄 때만 사용해보는 것이 적절한 책임을 찾는데에 도움이 될 수 있을 것 같아요!

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,61 @@
+		this.winningRank = Rank.MISS;
+	}
+
+	public void recordeRank(LottoNumbers winningNumbers, LottoNumber bonusBallNumber) {
+		int count = countMatchedNumber(winningNumbers);
+		for (Rank rank : Rank.values()) {
+			confirmRank(count, rank);
+		}
+		confirmRankThirdOrSecond(count, isMatch(bonusBallNumber));
+	}
```
LottoNumbers와 LottoNumber를 상태로 가지는 WinningLotto 클래스 도출를 도출해보면 어떨까요 ?

파라미터 인자의 수도 줄고, 좀 더 가독성 좋은 코드가 될 수 있을 것 같기도해요 :)

함께 WinningLotto 클래스는 어떤 책임을 가질 수 있을까에 대해 생각해보시면 좋을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,61 @@
+
 public class Lotto {
 	private final LottoNumbers lottoNumbers;
+	private Rank winningRank;
```
Lotto가 반드시 Rank를 상태로 저장하고 있어야할 까요?
Rank를 메서드의 반환 값으로만 사용하는 방법도 있을 것 같고, 
Rank를 기록하는 책임을 더욱 적합한 클래스에게 이전해볼 수 도 있을 것 같아요 🤔

leeyohan93 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,53 @@
+public class LottoNumber {
+	public static final int MIN_BOUND = 1;
+	public static final int MAX_BOUND = 45;
+	private static final String ERROR_LOTTO_NUMBER_BOUNDARY = "[ERROR] 로또번호는 1~45의 값을 가져야 합니다.";
+	public static final String ERROR_LOTTO_NUMBER_VALUE = "[ERROR] 로또번호는 숫자만 입력가능합니다.";
```
에러 메시지는 따로 관리해주시는 것도 가독성에 좋습니다 ! :)
참고만 해주세요 😄

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,61 @@
+	private int countMatchedNumber(LottoNumbers winningNumbers) {
 		int count = 0;
-		for (Integer winningNumber : winningNumbers.value()) {
-			count += lottoNumbers.isMatch(winningNumber);
+		for (LottoNumber lottoNumber : winningNumbers.getLottoNumbers()) {
+			count += this.lottoNumbers.ifMatchCount(lottoNumber);
 		}
 		return count;
```
이 부분도 Stream을 사용해보시면 좋을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,53 @@
+	private static int stringValueToIntValue(String stringNumber) {
+		try {
+			return Integer.parseInt(stringNumber);
+		} catch (NumberFormatException exception) {
+			throw new IllegalArgumentException(ERROR_LOTTO_NUMBER_VALUE);
```
NumberFormatException 같은 경우 IllegalArgumentException을 상속하고 있는데요.
예외의 추상화 레벨을 올리신 이유가 있을까요 ?
아래 링크가 도움이 될 수도 있을 것 같아요 :)

> [이펙티브 자바 아이템 73. 추상화 수준에 맞는 예외를 던지라](https://madplay.github.io/post/effectivejava-chapter10-exceptions#%EC%95%84%EC%9D%B4%ED%85%9C-73-%EC%B6%94%EC%83%81%ED%99%94-%EC%88%98%EC%A4%80%EC%97%90-%EB%A7%9E%EB%8A%94-%EC%98%88%EC%99%B8%EB%A5%BC-%EB%8D%98%EC%A7%80%EB%9D%BC)

grow-with replied:
해당 부분은 그 문자열로 입력받은 값을 내부에서 사용하기 위해 숫자로 파싱 할 때 , 숫자가 아닌 값이 들어오면 에러를 발생시키는 코드입니다.
제가 의도한 바는 NumberFormatException보다 IllegalArgumentException이 좀 더 잘못된 값을 입력했다는 뉘앙스를 주고 있고, 또한 에러의 경우마다 전달하고자 하는 메세지를 달리 하기 위함이였습니다.
제가 조금 더 깊게 생각해봐야 할 부분이 있을까요??

leeyohan93 commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -3,37 +3,70 @@
 	}
+
+	private void validateLottoNumberLength(List<LottoNumber> lottoNumbers) {
+		if (lottoNumbers.size() < MAX_RANGE_VALUE) {
```
로또 번호가 6개 미만이면 검증에 실패하지 않을까요?
아래와 같이 수정해보는건 어떨까요?
```suggestion
		if (lottoNumbers.size() == LOTTO_NUMBER_COUNT) {
```

grow-with replied:
안녕하세요 새벽에 자꾸 질문을 드리게 되네요 죄송합니다. ㅠ
그 해당 리뷰해주신부분은 로또번호가 6개 미만으로 검출되면 에러가 발생하도록 작성한 코드입니다.
그래서 6개 미만일경우에 에러가 발생하는게 맞는데 수정해주신 코드로 바꾸게 되면 6개 정상적으로 생성되었을때 에러가 발생하는 것 같습니다.
혹시 제가 리뷰해주신 부분에서 잘 못 이해하고 있는부분이 있을까요???

leeyohan93 replied:
제가 반대로 이야기드렸네요..  😅  죄송합니다 🙇 
6개 이상일때를 이야기드린 것이었습니다 :)
로또 번호는 항상 6개로 고정되어있어야 하지 않을까요? 🤔

grow-with replied:
아하 넵 수정했습니다!

leeyohan93 commented on src/main/java/lotto/domain/LottoNumbers.java:
```diff
@@ -3,37 +3,70 @@
+
+	private static void validateDuplicate(String[] stringNumberArray) {
+		Set<String> duplicateCheck = new HashSet<>();
+		for (String number : stringNumberArray) {
+			duplicateCheck.add(number);
+		}
```
Collections API를 사용해보는 건 어떨까요?

```suggestion
		Collections.addAll(duplicateCheck, stringNumberArray);
```

leeyohan93 commented on src/main/java/lotto/controller/LottoApplicationController.java:
```diff
@@ -4,61 +4,42 @@
-			lottoStatistics.record(lotto.countMatchNumber(winningNumbers));
+	public List<Lotto> recorde(String winningNumbers, int bonusBallNumber) {
+		for (Lotto lotto : lottos.getLottos()) {
+			lotto.recordeRank(new LottoNumbers(winningNumbers), new LottoNumber(bonusBallNumber));
```
로또의 생성에 대한 책임을 가지는 LottoShop 클래스를 도출했지만, 아직 생성에 대한 책임이 모두 이전된 것 같지 않네요 🤔 
책임을 모아주는건 어떨까요?

grow-with replied:
음. ㅠ 해당 부분은 로또를 생성한다기보다는 LottoNumbers라는 로또번호들을 의미하는 클래스와 LottoNumber 하나으 로또번호를 의미하는 클래스를 파라미터로 던져준 부분인데요.
그 저번에 주신 피드백에서 당첨번호도 결국 로또번호니까 LottoNumbers로 사용하시는게 좋을것 같다라는 피드백이 있으셨고, 보너스 번호도 결국 하나의 로또번호라서 LottoNumber 객체를 생성해서 파라미터로 사용하였습니다.
말씀해주신 로또 생성이라는 책임이 LottoShop에서 Lotto이든 LottoNumbers나 LottoNumber와 같이 로또 번호와 관련된건 전부 관리해야 한다는 의미로 생각해야할까요??

leeyohan93 replied:
> 로또 생성이라는 책임이 LottoShop에서 Lotto이든 LottoNumbers나 LottoNumber와 같이 로또 번호와 관련된건 전부 관리해야 한다는 의미로 생각해야할까요??

개인적으로는 로또를 생성하는 책임을 가지는 클래스를 별도로 정의했기 때문에 
LottoShop에 모아주는 것이 일관성 있고 유지보수하기 좋은 설계라고 생각합니다 😄 
만약, LottoShop에서 로또 번호와 관련된 생성의 책임을 담당해주고 Lotto를 반환해준다면 
그 이후의 객체들은 Lotto와 협력하면 되고 외부에서 LottoNumber, LottoNumbers에 대한 의존성을 제거할 수 있다는 장점이 있습니다. :)

leeyohan93 reviewed (changes requested):
이번 단계에서도 많이 고민하신 흔적이 보입니다! 🔥 
구조적인 개선이 쉽지 않았을 텐데 잘 반영해주셨어요! 👍 
몇가지 코멘트들을 남겨드렸으니 확인 부탁드릴게요 :)
궁금한 부분이 있으시다면 언제든 DM이나 코멘트 남겨주세요 😄

grow-with committed:
refactor: Repeater 패키지 이동, LOTTO_PRICE 상수 클래스 이동

grow-with committed:
refactor: Money 클래스 생성 및 테스트코드 작성

grow-with committed:
refactor: WinningLotto 생성, console 패키지 생성

grow-with committed:
refactor: 피드백에 따라서 전체 수정

- 책임 분리에 의해 controller 클래스 제거
- Lotto 클래스 내에 당첨등수 저장 변수 제거
- 당첨자를 구하는 메서드 Rank 클래스에 이동
- LottoShop의 sell()메서드에 금액 인자로 받도록 수정
- 전체 수정에 의한 테스트코드 수정

grow-with committed:
refactor: LottoNumbers 검증로직 6자리만 체크하도록 수정

leeyohan93 commented on src/main/java/lotto/domain/LottoShop.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.stream.IntStream;
+
+public class LottoShop {
+	public static final int LOTTO_PRICE = 1000;
```
https://github.com/next-step/java-lotto-pro/pull/241#discussion_r745674279
금액을 의미하는 상수 또한 Money 객체로 래핑할 수 있을 것 같아요 :)

```suggestion
	public static final Money LOTTO_PRICE = new Money(1000);
```

leeyohan93 commented on src/main/java/lotto/domain/LottoShop.java:
```diff
@@ -0,0 +1,36 @@
+		Collections.shuffle(allLottoNumbers);
+		List<LottoNumber> lottoNumbers = allLottoNumbers.subList(LottoNumbers.MIN_RANGE_VALUE,
+			LottoNumbers.MAX_RANGE_VALUE);
+		Collections.sort(lottoNumbers, (l1, l2) -> Integer.compare(l1.getValue(), l2.getValue()));
```
List의 sort를 쓰면 좀 더 가독성 있게 바꿔볼 수도 있을 것 같아요 :)

```suggestion
		lottoNumbers.sort(Comparator.comparingInt(LottoNumber::getValue));
```

leeyohan93 commented on src/main/java/lotto/domain/LottoShop.java:
```diff
@@ -0,0 +1,36 @@
+	}
+
+	public static Lottos sell(Money purchaseAmount) {
+		List<Lotto> lottos = new ArrayList<>();
+		IntStream.rangeClosed(1, purchaseAmount.getPurchaseQuantity(LOTTO_PRICE))
+			.forEach((i) -> lottos.add(create()));
+		return new Lottos(lottos);
```
IntStream에 대한 사용을 개선해보면 어떨까요? 해당 예시가 항상 정답은 아니니 참고만 해주세요 😄 
```suggestion
        return IntStream.range(0, purchaseAmount.getPurchaseQuantity(LOTTO_PRICE))
                .mapToObj(ignore -> create())
                .collect(Collectors.collectingAndThen(Collectors.toList(), Lottos::new));```

grow-with replied:
하나 배워갑니다!

leeyohan93 commented on src/main/java/lotto/domain/LottoShop.java:
```diff
@@ -0,0 +1,36 @@
+		return new Lotto(new LottoNumbers(lottoNumbers));
+	}
+
+	private static Lotto create(String lottoNumbers) {
+		return new Lotto(new LottoNumbers(lottoNumbers));
+	}
```
사용하지 않는 코드는 제거해주세요 :)

grow-with replied:
다음 5단계 수동생성을 위해 미리 만드부분인데 삭제할까요? ㅠ

leeyohan93 commented on src/main/java/lotto/domain/LottoShop.java:
```diff
@@ -0,0 +1,36 @@
+		return new Lotto(new LottoNumbers(lottoNumbers));
+	}
+
+	public static Lottos sell(Money purchaseAmount) {
```
sell에 대한 메서드를 개선해주셨네요 ! 💯

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,41 @@
 	}
 
-	public Lotto(int... lottoNumbers) {
+	public Lotto(final String lottoNumbers) {
```
해당 생성자도 사용하지 않고 있네요 :) 제거 부탁드릴게요

grow-with replied:
여기도 수동생성을 위해 미리 작성했어요 ㅠ

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,41 @@
+		return lottoNumbers.getLottoNumbers();
+	}
+
+	public Stream<LottoNumber> stream() {
+		return lottoNumbers.stream();
+	}
```
일급 컬렉션에서 stream을 반환하는 메서드를 만들기보다는 내부에서 처리될 수 있도록 하는 것이 좋을 것 같아요.
일급 컬렉션을 쓰는 이유 중 `상태와 행위를 한곳에서 관리` 에 대해 생각해보면 좋을 것 같습니다 :)
https://jojoldu.tistory.com/412

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,41 @@
+		this.lottoNumbers = lottoNumbers;
+	}
+
+	public long countMatchedNumber(WinningLotto winningLotto) {
+		return this.lottoNumbers.stream()
+			.filter((lottoNumber) -> winningLotto.stream()
+				.anyMatch(winningNumber -> winningNumber.equals(lottoNumber)))
+			.count();
 	}
```
현재 WinningLotto 클래스는 winningNumbers와 bonusBallNumber를 그저 담아두는 클래스로 사용되고 있는데요.
WinningLotto의 적절한 책임은 무엇일까요? 
아래 예시와 같이 WinningLotto를 통해 matchCount를 구할 수도 있고, 
근본적인 책임에 대해 생각해보면 우승 로또 (WinnigLotto)에 로또 1줄 (Lotto)를 매치하여 Rank를 반환하도록 해볼 수도 있을 것 같아요 :)

```suggestion
	public long countMatchedNumber(WinningLotto winningLotto) {
		return winningLotto.matchCount(lottoNumbers);
	}
```

leeyohan93 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -2,24 +2,60 @@
 
-	private final int matchedNumber;
+	private final int countOfMatch;
 	private final int winningAmount;
```
금액을 의미하는 winningAmount 도 Money 클래스로 만들어볼 수 있을 것 같아요 :)

```suggestion
	private final Money winningAmount;
```

leeyohan93 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -2,24 +2,60 @@
 		return winningAmount;
 	}
+
+	public int countWinners(Lottos lottos, WinningLotto winningLotto) {
+		int count = 0;
+		for (Lotto lotto : lottos.getLottos()) {
+			count += ifSameRankCount(lotto.countMatchedNumber(winningLotto),
+				lotto.isMatch(winningLotto.getBonusBallNumber()));
+		}
+		return count;
+	}
```
지금 로직에서는 countWinners를 반복적으로 호출하기 때문에, 성능적으로 좋지 않아보이는데요 🤔 
로또들을 한번만 순회해서 각 Rank별 count를 저장하는 LottoResult 클래스를 도출해보는건 어떨까요?

leeyohan93 reviewed (changes requested):
피드백 잘 반영해주셨습니다 ! 👍 
이번에도 몇가지 코멘트들을 남겨드렸으니 확인부탁드릴게요 :)
열정적인 모습 너무 좋습니다! 끝까지 화이팅해주세요 🔥

grow-with committed:
refactor: LottoPrice 상수 Money로 래핑

grow-with committed:
refactor: winningAmount 상수 Money로 래핑

grow-with committed:
refactor: 피드백에 따라서 코드 전체 수정

- 기존의 LottoNumbers 클래스 삭제
  6개의 번호를 의미하는데 이는 결국 로또 1장의 내용
- LottoResult 클래스를 통해 통계 저장 및 출력에 사용
- 기존의 당첨자 구하는 로직 수정

grow-with committed:
refactor: 안쓰는 코드 삭제

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,25 +1,74 @@
+		}
+	}
+
+	public long matchCount(Lotto lotto) {
+		return lottoNumbers.stream()
+			.filter(lottoNumber -> lotto.lottoNumbers
+				.stream()
+				.anyMatch(targetLottoNumber -> lottoNumber.equals(targetLottoNumber)))
+			.count();
+	}
```
List의 contains와 메서드참조를 사용하면 좀 더 가독성 있는 코드를 짤 수 있을 것 같아요 :)

```suggestion
	public long matchCount(Lotto lotto) {
		return lottoNumbers.stream()
			.filter(lottoNumbers::contains)
			.count();
	}
```

leeyohan93 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -1,25 +1,49 @@
+		return winningAmount.value();
+	}
+
+	public static Rank checkRank(long countOfMatch, boolean bonusNumberMatch) {
+		if (countOfMatch < Rank.FIFTH.getCountOfMatch()) {
+			return Rank.MISS;
+		}
+		if (countOfMatch == Rank.SECOND.getCountOfMatch()) {
+			return secondOrThirdRank(bonusNumberMatch);
+		}
+		return Arrays.stream(Rank.values())
+			.filter(rank -> rank.getCountOfMatch() == countOfMatch)
+			.findFirst()
+			.get();
+	}
```
이렇게 개선해볼 수도 있을 것 같아요 :)
```suggestion
	if (countOfMatch == Rank.SECOND.getCountOfMatch()) {
			return secondOrThirdRank(bonusNumberMatch);
		}
		return Arrays.stream(Rank.values())
			.filter(rank -> rank.getCountOfMatch() == countOfMatch)
			.findFirst()
			.orElse(MISS);
	}
```

leeyohan93 commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -0,0 +1,17 @@
+		this.bonusBallNumber = new LottoNumber(bonusBallNumber);
+	}
+
+	public Rank getRank(Lotto lotto) {
+		long countOfMatch = winningNumbers.matchCount(lotto);
+		boolean bonusNumberMatch = lotto.isMatch(bonusBallNumber);
+		return Rank.checkRank(countOfMatch, bonusNumberMatch);
+	}
```
새로 작성된 메서드에 대한 테스트와 몇몇 도메인 클래스에 대한 단위테스트를 깜빡하신 것 같아요 :)
마지막 단계를 진행하기 이전에 테스트들을 작성해주세요 👍 
테스트를 먼저 작성하는 것은 쉽지 않지만. 해당 단계는 TDD를 배우는 미션이기 때문에 마지막 요구사항은 TDD로 구현해보도록 노력해주세요 :)

leeyohan93 reviewed (approved):
피드백 잘 반영해주셨네요 !
눈에 띄게 개선된 점들이 많습니다 ! 👍 👍 
몇가지 코멘트를 남겨 놓았으니 확인 부탁드리고
이번 단계는 머지하겠습니다 :)
다음 단계가 마지막 단계인데요, 마지막까지 화이팅입니다 ! 🔥
